crates/driver/src/domain/competition/settled/mod.rs

```diff
@@ -0,0 +1,275 @@
+                    .amount
+                    .0
+                    .checked_mul(self.executed.into())?
+                    .checked_div(self.buy.amount.into())?;
+                // difference between limit sell and executed amount converted to sell token
+                limit_sell.checked_sub(
+                    self.executed
+                        .0
+                        .checked_mul(self.custom_price.buy)?
+                        .checked_div(self.custom_price.sell)?,
+                )
+            }
+            Side::Sell => {
+                // scale limit buy to support partially fillable orders
+                let limit_buy = self
+                    .executed
+                    .0
+                    .checked_mul(self.buy.amount.into())?
+                    .checked_div(self.sell.amount.into())?;
+                // difference between executed amount converted to buy token and limit buy
+                self.executed
+                    .0
+                    .checked_mul(self.custom_price.sell)?
+                    .checked_div(self.custom_price.buy)?
+                    .checked_sub(limit_buy)
+            }
+        }
+        .map(|surplus| eth::Asset {
+            token: self.surplus_token(),
+            amount: surplus.into(),
+        })
+    }
+
+    /// Surplus based on custom clearing prices returns the surplus after all
+    /// fees have been applied.
+    ///
+    /// Denominated in NATIVE token
+    fn native_surplus(
+        &self,
+        prices: &auction::NormalizedPrices,
+    ) -> Result<eth::TokenAmount, Error> {
+        let surplus = self
+            .surplus()
+            .ok_or(Error::Surplus(self.sell, self.buy))?
+            .amount;
+        let native_price = self.surplus_token_price(prices)?;
+        big_rational_to_u256(&(surplus.0.to_big_rational() * native_price.0))
+            .map(Into::into)
+            .map_err(Into::into)
+    }
+
+    /// Protocol fee is defined by fee policies attached to the order.
+    ///
+    /// Denominated in SURPLUS token
+    fn protocol_fee(&self) -> Result<eth::Asset, Error> {
+        // TODO: support multiple fee policies
+        if self.policies.len() > 1 {
+            return Err(Error::MultipleFeePolicies);
+        }
+
+        let protocol_fee = |policy: &order::FeePolicy| {
+            match policy {
+                order::FeePolicy::Surplus {
+                    factor,
+                    max_volume_factor,
+                } => Ok(std::cmp::min(
+                    {
+                        // If the surplus after all fees is X, then the original surplus before
+                        // protocol fee is X / (1 - factor)
+                        let surplus = self
+                            .surplus()
+                            .ok_or(Error::Surplus(self.sell, self.buy))?
+                            .amount;
+                        apply_factor(surplus.into(), factor / (1.0 - factor))
```

Comment:
Potential division by zero

Reply:
covered inside `apply` function where the f64::inf would be covered by conversion methods.